hera_cal/red_groups.py

```diff
@@ -0,0 +1,246 @@
+"""Module with classes and functions for dealing with redundant baseline groups."""
+from __future__ import annotations
+
+import attrs
+from .utils import reverse_bl
+from . import redcal
+import numpy as np
+from functools import cached_property
+
+from typing import Sequence
+
+AntPair = tuple[int, int]
+Baseline = tuple[int, int, str]
+
+def _convert_red_list(red_list: Sequence[Sequence[AntPair | Baseline]]) -> tuple[tuple[AntPair]]:
+    return [[r[:2] for r in red] for red in red_list]
+
+@attrs.define(frozen=False)
+class RedundantGroups:
+    """Redundant baseline group manager.
+
+    This class is used to manage redundant baseline groups. It is (roughly)
+    API-compatible with the list-of-lists-of-tuples format returned by, e.g., 
+    :func:`hera_cal.redcal.get_reds`, and in the future that may simply return
+    one of these objects.
+
+    To keep this compatibility, this object is mutable, but we absolutely discourage
+    directly mutating it. Instead, use the methods provided, including the `.append`
+    method, which is able to add a new redundant group to the object. 
+
+    In the future, we may disable direct mutation of this object.
+
+    Parameters
+    ----------
+    antpos : dict[int, np.ndarray]
+        Dictionary mapping antenna number to antenna position in ENU frame.
+        Not required if ``red_list`` is provided, but highly encouraged even if so.
+    bl_error_tol : float
+        Baseline length error tolerance in meters.
+    include_autos : bool
+        Whether to include autocorrelations in the redundant baseline groups.
+    red_list : list[list[tuple[int, int]]]
+        List of redundant baseline groups. Each group is a list of antenna pairs.
```

Comment:
I think we ought to be moving away from antpairs wherever possible. It's an older access pattern that isn't used much and I'm fine restricting this class to only doing len-3 bl tuples. If someone really needs len-2 tuples, they can use a list of lists. But I'd like to push us away from that since it creates so many annoying special cases.

As it stands, does this support RedundantGroups with different lists of baselines per polarization? It's not clear to me.

Reply:
Actually, what this class does is natively store the redundancies as simply antpairs, because that's all that's needed. However, when doing things like `__getitem__` with a key, the key can _either_ be an antpair or len-3 baseline tuple, and the result will be in the same form. I felt like this was the cleanest way. 

The answer to your question is _no_, it doesn't support different lists of baselines per polarization, but I think that makes sense, because this class is not meant to represent the actual baselines in any given dataset, but rather the set of redundancies in an array.

Reply:
I gave in and let it natively use antpair-pol tuples (it can actually use either).